File: bindings/csharp/Satox/Transactions/TransactionManager.cs
using System;
using System.Runtime.InteropServices;
using Satox.Bindings.Transactions;

namespace Satox.Transactions
{
    /// <summary>
    /// TransactionManager provides transaction handling and validation functionality.
    /// </summary>
    public interface ITransactionManager : IDisposable
    {
        void Initialize();
        void Shutdown();
        Transaction CreateTransaction(string from, string to, decimal amount, decimal fee);
        Transaction SignTransaction(Transaction transaction, byte[] privateKey);
        bool ValidateTransaction(Transaction transaction);
        TransactionReceipt BroadcastTransaction(Transaction transaction);
        TransactionInfo GetTransactionByHash(string transactionHash);
        TransactionStatus GetTransactionStatus(string transactionHash);
    }

    public class TransactionManager : ITransactionManager
    {
        private readonly Satox.Bindings.Transactions.TransactionManager _manager;

        /// <summary>
        /// Creates a new instance of TransactionManager.
        /// </summary>
        public TransactionManager()
        {
            _manager = new Satox.Bindings.Transactions.TransactionManager();
        }

        /// <summary>
        /// Initializes the transaction manager.
        /// </summary>
        public void Initialize()
        {
            _manager.Initialize();
        }

        /// <summary>
        /// Shuts down the transaction manager.
        /// </summary>
        public void Shutdown()
        {
            _manager.Shutdown();
        }

        /// <summary>
        /// Creates a new transaction.
        /// </summary>
        /// <param name="from">The sender address.</param>
        /// <param name="to">The recipient address.</param>
        /// <param name="amount">The transaction amount.</param>
        /// <param name="fee">The transaction fee.</param>
        /// <returns>The created transaction.</returns>
        public Transaction CreateTransaction(string from, string to, decimal amount, decimal fee)
        {
            return _manager.CreateTransaction(from, to, amount, fee);
        }

        /// <summary>
        /// Signs a transaction.
        /// </summary>
        /// <param name="transaction">The transaction to sign.</param>
        /// <param name="privateKey">The private key to sign with.</param>
        /// <returns>The signed transaction.</returns>
        public Transaction SignTransaction(Transaction transaction, byte[] privateKey)
        {
            return _manager.SignTransaction(transaction, privateKey);
        }

        /// <summary>
        /// Validates a transaction.
        /// </summary>
        /// <param name="transaction">The transaction to validate.</param>
        /// <returns>True if the transaction is valid, false otherwise.</returns>
        public bool ValidateTransaction(Transaction transaction)
        {
            return _manager.ValidateTransaction(transaction);
        }

        /// <summary>
        /// Broadcasts a transaction to the network.
        /// </summary>
        /// <param name="transaction">The transaction to broadcast.</param>
        /// <returns>The transaction receipt.</returns>
        public TransactionReceipt BroadcastTransaction(Transaction transaction)
        {
            return _manager.BroadcastTransaction(transaction);
        }

        /// <summary>
        /// Gets a transaction by its hash.
        /// </summary>
        /// <param name="transactionHash">The transaction hash.</param>
        /// <returns>The transaction information.</returns>
        public TransactionInfo GetTransactionByHash(string transactionHash)
        {
            return _manager.GetTransactionByHash(transactionHash);
        }

        /// <summary>
        /// Gets the transaction status.
        /// </summary>
        /// <param name="transactionHash">The transaction hash.</param>
        /// <returns>The transaction status.</returns>
        public TransactionStatus GetTransactionStatus(string transactionHash)
        {
            return _manager.GetTransactionStatus(transactionHash);
        }

        public void Dispose()
        {
            _manager.Dispose();
        }
    }

    /// <summary>
    /// Represents a transaction.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public struct Transaction
    {
        /// <summary>
        /// The transaction hash.
        /// </summary>
        [MarshalAs(UnmanagedType.LPStr)]
        public string Hash;

        /// <summary>
        /// The sender address.
        /// </summary>
        [MarshalAs(UnmanagedType.LPStr)]
        public string From;

        /// <summary>
        /// The recipient address.
        /// </summary>
        [MarshalAs(UnmanagedType.LPStr)]
        public string To;

        /// <summary>
        /// The transaction amount.
        /// </summary>
        public double Amount;

        /// <summary>
        /// The transaction fee.
        /// </summary>
        public double Fee;

        /// <summary>
        /// The transaction timestamp.
        /// </summary>
        public long Timestamp;

        /// <summary>
        /// The transaction signature.
        /// </summary>
        public IntPtr Signature;

        /// <summary>
        /// The length of the signature.
        /// </summary>
        public int SignatureLength;
    }

    /// <summary>
    /// Represents transaction information.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public struct TransactionInfo
    {
        /// <summary>
        /// The transaction hash.
        /// </summary>
        [MarshalAs(UnmanagedType.LPStr)]
        public string Hash;

        /// <summary>
        /// The block hash containing this transaction.
        /// </summary>
        [MarshalAs(UnmanagedType.LPStr)]
        public string BlockHash;

        /// <summary>
        /// The transaction timestamp.
        /// </summary>
        public long Timestamp;

        /// <summary>
        /// The transaction amount.
        /// </summary>
        public double Amount;

        /// <summary>
        /// The transaction fee.
        /// </summary>
        public double Fee;

        /// <summary>
        /// The sender address.
        /// </summary>
        [MarshalAs(UnmanagedType.LPStr)]
        public string From;

        /// <summary>
        /// The recipient address.
        /// </summary>
        [MarshalAs(UnmanagedType.LPStr)]
        public string To;

        /// <summary>
        /// The transaction status.
        /// </summary>
        public int Status;
    }

    /// <summary>
    /// Represents a transaction receipt.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public struct TransactionReceipt
    {
        /// <summary>
        /// The transaction hash.
        /// </summary>
        [MarshalAs(UnmanagedType.LPStr)]
        public string Hash;

        /// <summary>
        /// The block hash containing this transaction.
        /// </summary>
        [MarshalAs(UnmanagedType.LPStr)]
        public string BlockHash;

        /// <summary>
        /// The block number.
        /// </summary>
        public long BlockNumber;

        /// <summary>
        /// The transaction status.
        /// </summary>
        public int Status;

        /// <summary>
        /// The gas used.
        /// </summary>
        public long GasUsed;

        /// <summary>
        /// The cumulative gas used.
        /// </summary>
        public long CumulativeGasUsed;

        /// <summary>
        /// The contract address (if any).
        /// </summary>
        [MarshalAs(UnmanagedType.LPStr)]
        public string ContractAddress;
    }

    /// <summary>
    /// Represents the transaction status.
    /// </summary>
    public enum TransactionStatus
    {
        /// <summary>
        /// The transaction is pending.
        /// </summary>
        Pending = 0,

        /// <summary>
        /// The transaction is confirmed.
        /// </summary>
        Confirmed = 1,

        /// <summary>
        /// The transaction failed.
        /// </summary>
        Failed = 2,

        /// <summary>
        /// The transaction is unknown.
        /// </summary>
        Unknown = 3
    }
} 

File: bindings/go/satox/transaction/transaction_manager.go
package transaction

import (
	"C"
	"errors"
	"unsafe"
)

// TransactionManager handles transaction operations
type TransactionManager struct {
	initialized bool
}

// NewTransactionManager creates a new TransactionManager instance
func NewTransactionManager() *TransactionManager {
	return &TransactionManager{
		initialized: false,
	}
}

// Initialize initializes the transaction manager
func (tm *TransactionManager) Initialize() error {
	if tm.initialized {
		return nil
	}

	result := C.satox_transaction_manager_initialize()
	if result != 0 {
		return errors.New("failed to initialize transaction manager")
	}

	tm.initialized = true
	return nil
}

// Shutdown shuts down the transaction manager
func (tm *TransactionManager) Shutdown() error {
	if !tm.initialized {
		return nil
	}

	result := C.satox_transaction_manager_shutdown()
	if result != 0 {
		return errors.New("failed to shutdown transaction manager")
	}

	tm.initialized = false
	return nil
}

// CreateTransaction creates a new transaction
func (tm *TransactionManager) CreateTransaction(config *TransactionConfig) (*Transaction, error) {
	tm.ensureInitialized()

	cConfig := C.struct_TransactionConfig{
		from_address: C.CString(config.FromAddress),
		to_address:   C.CString(config.ToAddress),
		amount:       C.int64_t(config.Amount),
		fee:          C.int64_t(config.Fee),
		data:         C.CString(config.Data),
	}
	defer func() {
		C.free(unsafe.Pointer(cConfig.from_address))
		C.free(unsafe.Pointer(cConfig.to_address))
		C.free(unsafe.Pointer(cConfig.data))
	}()

	var cTx C.struct_Transaction
	result := C.satox_transaction_manager_create_transaction(&cConfig, &cTx)
	if result != 0 {
		return nil, errors.New("failed to create transaction")
	}

	tx := &Transaction{
		Hash:        C.GoString(cTx.hash),
		FromAddress: C.GoString(cTx.from_address),
		ToAddress:   C.GoString(cTx.to_address),
		Amount:      int64(cTx.amount),
		Fee:         int64(cTx.fee),
		Data:        C.GoString(cTx.data),
		Timestamp:   int64(cTx.timestamp),
		Status:      C.GoString(cTx.status),
	}

	return tx, nil
}

// SignTransaction signs a transaction
func (tm *TransactionManager) SignTransaction(tx *Transaction, privateKey string) error {
	tm.ensureInitialized()

	cHash := C.CString(tx.Hash)
	cPrivateKey := C.CString(privateKey)
	defer func() {
		C.free(unsafe.Pointer(cHash))
		C.free(unsafe.Pointer(cPrivateKey))
	}()

	result := C.satox_transaction_manager_sign_transaction(cHash, cPrivateKey)
	if result != 0 {
		return errors.New("failed to sign transaction")
	}

	return nil
}

// BroadcastTransaction broadcasts a transaction to the network
func (tm *TransactionManager) BroadcastTransaction(tx *Transaction) error {
	tm.ensureInitialized()

	cHash := C.CString(tx.Hash)
	defer C.free(unsafe.Pointer(cHash))

	result := C.satox_transaction_manager_broadcast_transaction(cHash)
	if result != 0 {
		return errors.New("failed to broadcast transaction")
	}

	return nil
}

// GetTransaction gets a transaction by its hash
func (tm *TransactionManager) GetTransaction(hash string) (*Transaction, error) {
	tm.ensureInitialized()

	cHash := C.CString(hash)
	defer C.free(unsafe.Pointer(cHash))

	var cTx C.struct_Transaction
	result := C.satox_transaction_manager_get_transaction(cHash, &cTx)
	if result != 0 {
		return nil, errors.New("failed to get transaction")
	}

	tx := &Transaction{
		Hash:        C.GoString(cTx.hash),
		FromAddress: C.GoString(cTx.from_address),
		ToAddress:   C.GoString(cTx.to_address),
		Amount:      int64(cTx.amount),
		Fee:         int64(cTx.fee),
		Data:        C.GoString(cTx.data),
		Timestamp:   int64(cTx.timestamp),
		Status:      C.GoString(cTx.status),
	}

	return tx, nil
}

// GetTransactionStatus gets the status of a transaction
func (tm *TransactionManager) GetTransactionStatus(hash string) (string, error) {
	tm.ensureInitialized()

	cHash := C.CString(hash)
	defer C.free(unsafe.Pointer(cHash))

	var cStatus *C.char
	result := C.satox_transaction_manager_get_transaction_status(cHash, &cStatus)
	if result != 0 {
		return "", errors.New("failed to get transaction status")
	}
	defer C.free(unsafe.Pointer(cStatus))

	return C.GoString(cStatus), nil
}

func (tm *TransactionManager) ensureInitialized() {
	if !tm.initialized {
		panic("transaction manager not initialized")
	}
}

// TransactionConfig represents transaction configuration
type TransactionConfig struct {
	FromAddress string
	ToAddress   string
	Amount      int64
	Fee         int64
	Data        string
}

// Transaction represents a blockchain transaction
type Transaction struct {
	Hash        string
	FromAddress string
	ToAddress   string
	Amount      int64
	Fee         int64
	Data        string
	Timestamp   int64
	Status      string
}
